Source/Dafny/AST/DafnyAst.cs

```diff
@@ -4000,52 +3996,53 @@ public class ModuleDefinition : INamedRegion, IAttributeBearingDeclaration {
     }
 
     VisibilityScope visibilityScope;
+    public VisibilityScope VisibilityScope =>
+      visibilityScope ??= new VisibilityScope(this.SanitizedName);
 
-    public VisibilityScope VisibilityScope {
+    public virtual bool IsDefaultModule {
       get {
-        if (visibilityScope == null) {
-          visibilityScope = new VisibilityScope(this.CompileName);
-        }
-        return visibilityScope;
+        return false;
       }
     }
 
-    public virtual bool IsDefaultModule {
+    private string sanitizedName = null;
+
+    public string SanitizedName {
       get {
-        return false;
+        if (sanitizedName == null) {
+          if (IsBuiltinName) {
+            sanitizedName = Name;
+          } else if (EnclosingModule != null && EnclosingModule.Name != "_module") {
+            // Include all names in the module tree path, to disambiguate when compiling
+            // a flat list of modules.
+            // Use an "underscore-escaped" character as a module name separator, since
+            // underscores are already used as escape characters in SanitizeName()
+            sanitizedName = EnclosingModule.SanitizedName + "_m" + NonglobalVariable.SanitizeName(Name);
+          } else {
+            sanitizedName = NonglobalVariable.SanitizeName(Name);
+          }
+        }
+        return sanitizedName;
       }
     }
+
     string compileName;
     public string CompileName {
       get {
         if (compileName == null) {
           var externArgs = DafnyOptions.O.DisallowExterns ? null : Attributes.FindExpressions(this.Attributes, "extern");
           if (externArgs != null && 1 <= externArgs.Count && externArgs[0] is StringLiteralExpr) {
             compileName = (string)((StringLiteralExpr)externArgs[0]).Value;
-          } else if (IsBuiltinName || externArgs != null) {
+          } else if (externArgs != null) {
```

Comment:
This case should only happen when `externArgs.Count == 0` right? Maybe be nice to make that explicit.

Reply:
Also when externArgs[1] is something else than a string, I think?  Though that should probably not be accepted.

Reply:
> Also when externArgs[1] is something else than a string, I think? Though that should probably not be accepted.

I'll leave it up to you to decide whether to make that more strict :-) From the documentation it doesn't look like a non-string argument is accepted.

Reply:
We can fix this separately, I think; I prefer for this PR to not change existing behaviors.

Reply:
> We can fix this separately, I think; I prefer for this PR to not change existing behaviors.

That's fair